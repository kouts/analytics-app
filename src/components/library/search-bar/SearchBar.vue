<template>
  <div class="search-bar-wrapper">
    <Icon name="search" class="search-bar-icon search-bar-icon--search" />
    <input
      ref="inputRef"
      v-model="searchText"
      type="search"
      class="search-bar-input"
      data-search-bar-input="true"
      v-bind="$attrs"
      autocomplete="off"
      :placeholder="$attrs.placeholder || 'Search...'"
      v-on="$listeners"
      @input="input"
      @keydown="inputKeydown"
    />
    <div class="search-bar-icon search-bar-icon--shortcut" title='Press "/" to search'>/</div>
    <div v-show="searchText.length > 0" class="search-bar-icon search-bar-icon--clear" @mousedown="clear">
      <Icon name="x" />
    </div>
  </div>
</template>

<script>
import Icon from '../icon'

/*eslint-disable */
/**
 * The `SearchBar` component displays a search input with some additional features built-in.  
 * 
 * **Features:**  
 * - Includes a default placeholder text - customizable using the `placeholder` attribute.  
 * - The search input **expands** when it gains focus.  
 * - Displays an `x` icon on the right side of the search input, used for **clearing** the text when there's a value typed inside.  
 * - **Focus** on the search input at any time by pressing the `/` key on the keyboard.  
 * - The search text gets cleared by pressing the `esc` key when the search input has focus.
 * 
 * ***Important:*** It is advisable that you include the `SearchBar` component **only once** on each page.  
 * In case multiple `SearchBar` components are present, the first one being displayed will take focus precedence upon `/` keypress.
 */
/* eslint-enable */
export default {
  components: {
    Icon
  },
  inheritAttrs: false,
  emits: ['update'],
  data() {
    return {
      searchText: ''
    }
  },
  created() {
    window.document.addEventListener('keydown', this.documentKeydown)
  },
  beforeDestroy() {
    window.document.removeEventListener('keydown', this.documentKeydown)
  },
  methods: {
    input(e) {
      /**
       * Event that fires when the search text is updated
       */
      this.$emit('update', e.target.value)
    },
    clear() {
      this.searchText = ''
      this.$emit('update', '')
    },
    inputKeydown(e) {
      if (e.key === 'Escape') {
        this.clear()
        this.$refs.inputRef.blur()
      }
    },
    documentKeydown(e) {
      if (
        e.key === '/' &&
        e.target !== this.$refs.inputRef &&
        window.document.activeElement !== this.$refs.inputRef &&
        e.target instanceof HTMLInputElement === false &&
        e.target instanceof HTMLSelectElement === false &&
        e.target instanceof HTMLTextAreaElement === false
      ) {
        e.preventDefault()
        const allVisibleSearchInputs = [].slice.call(document.querySelectorAll('[data-search-bar-input]')).filter((el) => {
          return !!(el.offsetWidth || el.offsetHeight || el.getClientRects().length)
        })
        const elToFocus = allVisibleSearchInputs.length > 1 ? allVisibleSearchInputs[0] : this.$refs.inputRef

        elToFocus.focus()
        elToFocus.select()
      }
    }
  }
}
</script>

<style lang="scss" scoped>
$input-color: #16212c;
$input-background: #eff2f7;
$icon-color: darken($input-background, 30%);

.search-bar-wrapper {
  display: inline-block;
  position: relative;
}

.search-bar-icon {
  color: $icon-color;
  position: absolute;
  bottom: 0.45rem;
  &--search {
    left: 0.75rem;
  }
  &--shortcut {
    cursor: text;
    display: inline-block;
    right: 0.75rem;
    padding: 1px 5px;
    border: 1px solid lighten($icon-color, 14%);
    background-color: darken($input-background, 3%);
    border-radius: 3px;
    z-index: 50;
    font-family: 'Inter', system-ui, BlinkMacSystemFont, 'Segoe UI', Roboto, 'Helvetica Neue', Arial, sans-serif;
    line-height: 1.15rem;
    font-weight: 400;
    font-size: 14px;
  }
  &--clear {
    right: 0.75rem;
    cursor: pointer;
    z-index: 10;
  }
}

.search-bar-input {
  display: block;
  font-family: 'Inter', system-ui, BlinkMacSystemFont, 'Segoe UI', Roboto, 'Helvetica Neue', Arial, sans-serif;
  width: 250px;
  padding: 0.375rem 2rem;
  font-size: 14px;
  font-weight: 400;
  line-height: 1.5;
  color: #16212c;
  background-color: $input-background;
  border: 1px solid $input-background;
  border-radius: 0.35rem;
  transition: border-color 0.15s ease-in-out, width 0.25s;
  &:focus {
    background-color: lighten($input-background, 3%);
    border-color: darken($input-background, 7%);
    outline: 0;
    box-shadow: none;
    width: 300px;
    ~ .search-bar-icon--shortcut {
      display: none;
    }
  }
}

/* Fix the X appearing in search field on Chrome and IE */
input[type='search']::-ms-clear {
  display: none;
  width: 0;
  height: 0;
}
input[type='search']::-ms-reveal {
  display: none;
  width: 0;
  height: 0;
}

input[type='search']::-webkit-search-decoration,
input[type='search']::-webkit-search-cancel-button,
input[type='search']::-webkit-search-results-button,
input[type='search']::-webkit-search-results-decoration {
  display: none;
}
</style>
