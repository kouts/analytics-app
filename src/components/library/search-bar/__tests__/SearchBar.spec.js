import SearchBar from '../SearchBar.vue'
import { mount } from '@vue/test-utils'

describe('Search Bar', () => {
  let wrapper

  beforeEach(() => {
    wrapper = mount(SearchBar)
  })

  it('should be defined', () => {
    expect(SearchBar).toBeDefined()
  })

  it('should show fire an update event when text changes', async () => {
    const inputSearch = wrapper.find('input[type="search"]')

    await inputSearch.setValue('some search text')
    await wrapper.vm.$nextTick()
    expect(wrapper.emitted().update).toBeTruthy()
    expect(wrapper.emitted().update[0]).toEqual(['some search text'])
  })

  it('should show an "x" icon when it has focus and contains a value', async () => {
    const inputSearch = wrapper.find('input[type="search"]')

    await inputSearch.setValue('some search text')
    inputSearch.trigger('focus')
    await wrapper.vm.$nextTick()
    expect(wrapper.find('.search-bar-icon--clear').isVisible()).toBe(true)
  })

  it('should clear the text when clicking on the "x" icon', async () => {
    const inputSearch = wrapper.find('input[type="search"]')

    await inputSearch.setValue('some search text')
    inputSearch.trigger('focus')
    await wrapper.vm.$nextTick()
    wrapper.find('.search-bar-icon--clear').trigger('mousedown')
    await wrapper.vm.$nextTick()
    expect(wrapper.find('input[type="search"]').element.value).toBe('')
    expect(wrapper.vm.searchText).toBe('')
  })

  it('should have a custom placeholder text', async () => {
    const customWrapper = mount(SearchBar, {
      attrs: {
        placeholder: 'custom placeholder text'
      }
    })
    const inputSearch = customWrapper.find('input[type="search"]')

    expect(inputSearch.attributes().placeholder).toBe('custom placeholder text')
  })

  it('should get focus when the "/" key is pressed', async () => {
    const customWrapper = mount(SearchBar, {
      attachTo: document.body
    })

    customWrapper.trigger('keydown', {
      key: '/'
    })
    await customWrapper.vm.$nextTick()
    expect(customWrapper.find('input[type="search"]').element).toBe(document.activeElement)
  })

  describe('snapshots', () => {
    it('should render a search input', async () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })
})
