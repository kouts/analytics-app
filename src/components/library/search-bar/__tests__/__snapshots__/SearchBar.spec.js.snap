// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Bar snapshots should render a search input 1`] = `
<div
  class="search-bar-wrapper"
>
  <div
    class="icon search-bar-icon search-bar-icon--search"
  >
    <svg
      aria-hidden="true"
      class="octicon octicon-search"
      height="16"
      version="1.1"
      viewBox="0 0 16 16"
      width="16"
    >
      <path
        d="M11.5 7a4.499 4.499 0 11-8.998 0A4.499 4.499 0 0111.5 7zm-.82 4.74a6 6 0 111.06-1.06l3.04 3.04a.75.75 0 11-1.06 1.06l-3.04-3.04z"
        fill-rule="evenodd"
      />
    </svg>
  </div>
   
  <input
    autocomplete="off"
    class="search-bar-input"
    data-search-bar-input="true"
    placeholder="Search..."
    type="search"
  />
   
  <div
    class="search-bar-icon search-bar-icon--shortcut"
    title="Press \\"/\\" to search"
  >
    /
  </div>
   
  <div
    class="search-bar-icon search-bar-icon--clear"
    style="display: none;"
  >
    <div
      class="icon"
    >
      <svg
        aria-hidden="true"
        class="octicon octicon-x"
        height="16"
        version="1.1"
        viewBox="0 0 16 16"
        width="16"
      >
        <path
          d="M3.72 3.72a.75.75 0 011.06 0L8 6.94l3.22-3.22a.75.75 0 111.06 1.06L9.06 8l3.22 3.22a.75.75 0 11-1.06 1.06L8 9.06l-3.22 3.22a.75.75 0 01-1.06-1.06L6.94 8 3.72 4.78a.75.75 0 010-1.06z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</div>
`;
